dashboard/modules/job/job_supervisor.py

```diff
@@ -0,0 +1,442 @@
+from ray._private.runtime_env.constants import RAY_JOB_CONFIG_JSON_ENV_VAR
+from ray.actor import ActorHandle
+from ray.dashboard.modules.job.common import (
+    JOB_ID_METADATA_KEY,
+    JOB_NAME_METADATA_KEY,
+    JobInfoStorageClient,
+)
+from ray.dashboard.modules.job.job_log_storage_client import JobLogStorageClient
+from ray.job_submission import JobStatus
+
+# asyncio python version compatibility
+try:
+    create_task = asyncio.create_task
+except AttributeError:
+    create_task = asyncio.ensure_future
+
+# Windows requires additional packages for proper process control.
+if sys.platform == "win32":
+    try:
+        import win32api
+        import win32con
+        import win32job
+    except (ModuleNotFoundError, ImportError) as e:
+        win32api = None
+        win32con = None
+        win32job = None
+        logger = logging.getLogger(__name__)
+        logger.warning(
+            "Failed to Import win32api. For best usage experience run "
+            f"'conda install pywin32'. Import error: {e}"
+        )
+
+
+class JobSupervisor:
+    """
+    Ray actor created by JobManager for each submitted job, responsible to
+    setup runtime_env, execute given shell command in subprocess, update job
+    status, persist job logs and manage subprocess group cleaning.
+
+    One job supervisor actor maps to one subprocess, for one job_id.
+    Job supervisor actor should fate share with subprocess it created.
+    """
+
+    DEFAULT_RAY_JOB_STOP_WAIT_TIME_S = 3
+    SUBPROCESS_POLL_PERIOD_S = 0.1
+    VALID_STOP_SIGNALS = ["SIGINT", "SIGTERM"]
+
+    def __init__(
+        self,
+        job_id: str,
+        entrypoint: str,
+        user_metadata: Dict[str, str],
+        gcs_address: str,
+    ):
+        self._job_id = job_id
+        gcs_aio_client = GcsAioClient(address=gcs_address)
+        self._job_info_client = JobInfoStorageClient(gcs_aio_client)
+        self._log_client = JobLogStorageClient()
+        self._entrypoint = entrypoint
+
+        # Default metadata if not passed by the user.
+        self._metadata = {JOB_ID_METADATA_KEY: job_id, JOB_NAME_METADATA_KEY: job_id}
+        self._metadata.update(user_metadata)
+
+        # fire and forget call from outer job manager to this actor
+        self._stop_event = asyncio.Event()
+
+        # Windows Job Object used to handle stopping the child processes.
+        self._win32_job_object = None
+
+        # Logger object to persist JobSupervisor logs in separate file.
+        self._logger = logging.getLogger(f"{__name__}.supervisor-{job_id}")
+        supervisor_log_file_name = os.path.join(
+            ray._private.worker._global_node.get_logs_dir_path(),
+            f"jobs/supervisor-{job_id}.log",
+        )
+        os.makedirs(os.path.dirname(supervisor_log_file_name), exist_ok=True)
+        self._logger.addHandler(logging.FileHandler(supervisor_log_file_name))
+        self._logger.addHandler(logging.StreamHandler())
```

Comment:
let's move this logic to a utility function or method like `configure_logger` instead of inlining the logic in the constructor. the constructor will quickly get out of hand this way, better to isolate logically separate logic